packages/js/email-editor/src/components/preview/preview-save-guard.ts

```diff
@@ -0,0 +1,102 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useCallback, useEffect } from '@wordpress/element';
+import { select, dispatch } from '@wordpress/data';
+import { store as editorStore } from '@wordpress/editor';
+import { store as noticesStore } from '@wordpress/notices';
+
+/**
+ * A component that adds a confirmation dialog for unsaved changes
+ * when a user tries to open a preview in a new tab.
+ */
+export const PreviewSaveGuard = () => {
+	const selector = '.editor-preview-dropdown__button-external';
+
+	/**
+	 * Handles click/keydown events to check for unsaved changes before previewing.
+	 *
+	 * @param {Event} event The triggered event.
+	 */
+	const guard = async ( event ) => {
+		const target = event.target;
+
+		// Find the triggering element (the preview button).
+		const triggerEl = target?.closest( selector );
+		if ( ! triggerEl ) {
+			return;
+		}
+
+		// Check if there are any unsaved changes using the editor data store.
+		const editorStoreInstance = select( editorStore );
+		const isDirty = editorStoreInstance?.isEditedPostDirty?.();
+
+		if ( ! isDirty ) {
+			// If the post is saved, do nothing and let the preview open normally.
+			return;
+		}
+
+		// If there are unsaved changes, prevent the default action (opening the link).
+		event.preventDefault();
+		event.stopPropagation();
+		event.stopImmediatePropagation();
+
+		// Dispatch a notice to inform the user they need to save first.
+		dispatch( noticesStore ).createNotice(
+			'info',
+			__(
+				'You have unsaved changes. Please save the post before previewing.',
+				'woocommerce'
+			),
+			{
+				type: 'snackbar',
+				isDismissible: true,
+			}
+		);
+	};
+
+	/**
+	 * Handles keydown events for the preview button.
+	 *
+	 * @param {KeyboardEvent} event The triggered event.
+	 */
+	const keydownHandler = useCallback(
+		( event ) => {
+			try {
+				const target = event.target;
+				// Trigger on 'Enter' or 'Space' keys if the target is the preview button.
```

Comment:
I am not very happy about the number of code comments you added in this file. Some of them, like this one, are not even helpful and explain nothing to me.